Add setup for user and developer documentation

oruebel opened the pull request:
This PR adds the basic setup for creating custom pages for the user and developer documentation. Currently this looks as follows.


<img width="326" alt="Screen Shot 2024-08-21 at 12 51 31 AM" src="https://github.com/user-attachments/assets/f9384d70-e0b6-4fdf-9b23-0ad57e4d44d5">


# CHANGES
- **Basic Setup**
    - [X] Added setup for creating User Docs and Developer Docs sections
    - [X] Added setup for creating testable code examples as part of the `tests/examples` folder. The code in the folder is being executed with the regular tests, but has additional markup to be able to include specific sections of the code in the docs. In this way the code snippets in the docs are being tests with the regular test suite. 
    - [X] Added `.codespell_ignore` file to ignore Doxygen commands when running `codespell`
- **New Documentation Pages**
    - [X] Moved requirements and installation instructions from the README to a dedicated page in the docs
    - [X] Moved LICENCE and COPYRIGHT pages from the README to the docs
    - [X] Added brief developer docs about testing
    - [X] Add docs for SWMR mode and chunking in HDF5
    - [X] Add example for making dot figures inline as part of the HDF5IO docs
    - [X] Add documentation about how to write Docs (including code examples)
- **Source Code Changes**
    - [X] Added `BaseIO::flush()` and `HDF5IO::flush()`
    - [X] Fixed format conversion warnings in HDF5 tests
    - [X] Added missing docstrings for namespaces

# TODO (possibly for future PR)

- Expand the testing docs to also include other tests (e.g., sanitize) that a user should run
- Add developer dosc about how to edit documentation
- Add Getting Started page for users
- Add tutorial on how to integrate AqNWB with a data acquisition system
- Add docs on how to implement an alternate storage backend

oruebel commented:
@stephprince would you prefer we define the basic setup here first and then add actual content for the docs in a separate PR or would you like me to add content (e.g., to describe SWMR) in this PR, so that we one PR for documentation content?

stephprince commented:
@oruebel I'm ok adding content all in this PR, so we don't have placeholders and since the folder structure may change as you make the content.

oruebel commented:
@stephprince please take a look. The main setup, outline, and a few additional docs pages are there. Given that this PR is already getting pretty big, I think it may be useful to use separate PR's to add more content to the docs.

stephprince commented on docs/pages/userdocs/hdf5io.dox:
```diff
@@ -0,0 +1,202 @@
+/**
+ * @page hdf5io HDF5 I/O
+ *
+ * \section hdf5io_chunking Chunking
```
This looks good but we currently don't have a way for the user to specify chunk size. It's defined in the `NWBFile.cpp` file. It's on the issue ticket list but not sure if we want to include it in the docs yet?

oruebel replied:
I think it is fine to include here since you are using chunking even if a user cannot configure this yet.

stephprince commented on docs/pages/userdocs/hdf5io.dox:
```diff
@@ -0,0 +1,202 @@
+ *
+ *
+ *
+ * \dot
```
All the diagrams look great!

stephprince reviewed (changes requested):
All the documentation and the diagrams look great! Added my suggestions.

oruebel commented on docs/pages/userdocs/install.dox:
```diff
@@ -0,0 +1,130 @@
+ *         "outputOnFailure": true
+ *       }
+ *     }
+ *   ]
+ * }
+ * \endcode
+ * Replace `<os>` in the `CMakeUserPresets.json` file with the name of
+ * the operating system you have (`win64`, `linux` or `darwin`).
+ *
+ * \subsection configure_build_test Configure, build and test
+ *
+ * You can configure, build and test the project respectively with the following commands from the project root on
+ * any operating system with any build system:
+ *
+ * \code{.sh}
+ * cmake --preset=dev
+ * cmake --build --preset=dev
+ * ctest --preset=dev
+ * \endcode
+ *
+ * \subsection devbuild_dev_mode_targets_subsec Developer mode targets
+ *
+ * Additional targets can be invoked when in development mode using the commands below
+ *
+ * \code{.sh}
+ * cmake --build --preset=dev --target=<name of the target>
+ * \endcode
+ *
+ * \subsubsection  devbuild_target_options_subsubsec Target options
+ * - `format-check`: run the `clang-format` tool on the codebase to check for formatting errors
+ * - `format-fix` :  run the `clang-format` tool on the codebase with `FIX=YES` to both check and automatically fix for formatting errors
+ * - `spell-check`: run the `codespell` tool on the codebase to check for common spelling errors
+ * - `spell-fix` :  run the `codespell` tool on the codebase with `FIX=YES` to both check and automatically fix  common spelling errors
+ * - `docs` : builds the documentation using Doxygen. (Note: run `cmake --preset=dev -DBUILD_DOCS=ON` before building to add docs target)
+ */
+
+
+
+
+ */
```
The way I considered it, a User is someone using AqNWB in their code (e.g.,to integrate NWB with their system) and a developer is someone contributing to AqNWB. Since we are in C++, both users and developers need to know how to build the library (or user at least need to be able to link to it if we do binary releases). We could split the page into two parts to have separate install pages in the user and developer sections.

oruebel replied:
I have separated the installation instructions for the users and developers on respective pages in the userdocs/devdocs as you suggested. I've moved the requirements back to the main page so that we can reference them from both installation docs and avoid duplication.

oruebel commented on README.md:
```diff
@@ -8,187 +8,16 @@
 Our goal is to provide a lightweight API to integrate with existing acquisition systems.
 
-Please note, AqNWB is currently under development and should not yet be used in practice.
+Please see the [Documentation](https://neurodatawithoutborders.github.io/aqnwb/) for 
```
For the links in the README to work both in GitHub and the Doxygen docs, I think we'll need to link to the URL on https://neurodatawithoutborders.github.io/aqnwb/ directly, rather than being able to dynamically generate the link with Doxygen as otherwise GitHub won't know what those links are.

oruebel commented:
@stephprince can you take another look. I believe I've addressed all your comments. I also added a new page to describe some of the details about how to add content to the documentation.

oruebel commented on tests/CMakeLists.txt:
```diff
@@ -17,24 +17,42 @@ add_executable(aq-nwb_test
+    examples/test_example.cpp
+)
+
+# Ensure the aq-nwb_test target can include headers from the current directory and Catch2
```
```suggestion
# Ensure the aq-nwb_test target can include headers from the current directory
```

stephprince commented on docs/pages/devdocs/documentation.dox:
```diff
@@ -0,0 +1,117 @@
+ * First we need to define our code example by creating a new ``cpp`` file
+ * in the ``tests/examples`` folder. Here we use the file ``test_examples.cpp`` shown below.
+ * The file looks like a regular unit test with the addition of Doxygen comments of the form
+ * `// [example_hdf5io_code_snippet]`` to mark regions in the code that we want to include
+ * in the docs as code sinppets.
```
```suggestion
 * ``// [example_hdf5io_code_snippet]`` to mark regions in the code that we want to include
 * in the docs as code snippets.
```

stephprince commented on docs/pages/devdocs/documentation.dox:
```diff
@@ -0,0 +1,117 @@
+ *  object we use:
+ *
+ * \code{.sh}
+ *  \snippet tests/examples/test_example.cpp example_hdf5io_code_snippet``
```
```suggestion
 *  \snippet tests/examples/test_example.cpp example_hdf5io_code_snippet
```

stephprince commented on docs/pages/devdocs/documentation.dox:
```diff
@@ -0,0 +1,117 @@
+ *
+ * \section dev_docs_graphviz_sec Creating Custom Dot Graphs
+ *
+ * AqNWB uses  [Graphviz](https://graphviz.org/) to improve the quality of
```
```suggestion
 * AqNWB uses [Graphviz](https://graphviz.org/) to improve the quality of
```

stephprince commented on docs/pages/devdocs/documentation.dox:
```diff
@@ -0,0 +1,117 @@
+ * 2. Enable building the docs by setting the  cmake variable ``BUILD_DOCS=ON``, e.g.,
+ *    by calling ``cmake --preset=dev -DBUILD_DOCS=ON``
+ *
+ * 3. Build the documentation by running: ``cmake --build --preset=dev --target=docs``
```
```suggestion
 * 3. Build the documentation by running: ``cmake --build --preset=dev --target=docs``. The resulting documentation files can be found in your build directory ``build/dev/docs/html``.
```

stephprince reviewed (commented):
A few small changes to section about adding new documentation but otherwise looks good!

oruebel commented:
> A few small changes to the section about adding new documentation but otherwise looks good!

Thanks for the additional comments. I committed all your suggestions.
